src/librustc_mir/transform/const_prop.rs

```diff
@@ -650,11 +650,25 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     // Rvalue is also a pair with two scalars. The more general case is more
                     // complicated to implement so we'll do it later.
                     // FIXME: implement the general case stated above ^.
+                    // Note that field reordering must be handled:
+                    //   `(u8, u128)` is represented as ScalarPair(u128, u8)
+                    // and ZST fields should be handled:
+                    //   `(u8, (), u128, ())` is represented as ScalarPair(u128, u8)
                     let ty = &value.layout.ty.kind;
+                    let shape = &value.layout.layout.fields;
                     // Only do it for tuples
-                    if let ty::Tuple(substs) = ty {
+                    if let (ty::Tuple(substs), FieldsShape::Arbitrary { memory_index, .. }) =
+                        (ty, shape)
+                    {
                         // Only do it if tuple is also a pair with two scalars
                         if substs.len() == 2 {
+                            // The components of a scalar pair are in memory order, so to generate
+                            // a tuple with the correct types we must put them back in source order.
+                            let (one, two) = match memory_index.as_slice() {
+                                [0, 1] => (one, two),
+                                [1, 0] => (two, one),
+                                _ => bug!("bad memory_index for 2-tuple: {:?}", memory_index),
+                            };
```

Comment:
Instead of fixing this, I would've preferred to this block entirely. The way scalar pair const prop is currently implemented (by generating a const tuple and assuming it has the same layout) is really fragile, and any layout change could break this again. However this regresses ~5 mir-opt tests (don't have a diff handy).